### PR TITLE
Use measureme in self profiler

### PR DESCRIPTION
**Pull request by wesleywiser**
r? @michaelwoerister 

~Changes are still very rough.~

~I'm not sure what the right way to add the `measureme` dependency is. Currently it's just added with a relative path which Works On My Machine ™️.~

I'm also not sure what to do with the category data.

### CODE REVIEW

File: src/librustc/util/profiling.rs [hunk 1]
```diff
@@ -35,409 +33,136 @@ pub enum ProfilerEvent {
...
+            (EventKind::GenericActivity, "GenericActivity"),
+            (EventKind::IncrementalLoadResult, "IncrementalLoadResult"),
+            (EventKind::QueryBlocked, "QueryBlocked"),
+            (EventKind::QueryCacheHit, "QueryCacheHit"),
+        ];
+
+        for (kind, s) in event_kinds {
+            let string_id = StringId::reserved(*kind as u32);
+            profiler.alloc_string_with_reserved_id(string_id, *s);
+        }
```
Comment (michaelwoerister) [inline src/librustc/util/profiling.rs]: I think I would get rid of the `EventKind` enum and just use a regular `alloc_string` here, storing the `StringId`s in fields of the `SelfProfiler` struct. That should have about the same performance and doesn't rely on magic numbers.

**Comment by michaelwoerister** [comment]
Looks like a good start! Eventually I'd like to have query names instead of dep-node names here but I don't know how to exactly do that yet. 

Have you done performance tests on Windows? If the mmap-sink performance worse than the file-sink there, I think I'd like to get rid of the `memmap` dependency and implement the mmap-sink only on Unix via direct `libc` calls (we'll need some stuff from `libc` that `memmap` doesn't support anyway). On Windows we'd just use the file-sink.

For now, I think we should just ignore the category data. Later we could encode it in the `event_id` if we wanted to.

I think we can just publish `measureme` on crates.io and have lots of major version changes. The other option would be a git dependency. But that's something we should discuss with `wg-meta`.

**Comment by michaelwoerister** [comment]
For reference: https://rust-lang.zulipchat.com/#narrow/stream/185694-t-compiler.2Fwg-meta/topic/out-of-tree.20crates.20-.20crates.2Eio.20or.20git.20dependency

**Comment by wesleywiser** [comment]
> Looks like a good start! Eventually I'd like to have query names instead of dep-node names here but I don't know how to exactly do that yet.

Originally I passed `Query<'tcx>`s into the profiler methods but I couldn't figure out how to preregister the query names so I had to use a `FxHashSet` to keep track of which query names had been interned into the string table and which hadn't. Maybe the overhead of that could be reduced more but removing it in favor of using dep-node names and pre-registering the names made a noticeable difference in performance.

> Have you done performance tests on Windows?

I've been working on getting rustc to build on my Windows box. It's been a long time since I used it and my msvc toolchain was completely out of date. I think that's resolved now and I'm just waiting for LLVM to build. I should have the performance data tonight.

> For now, I think we should just ignore the category data.

Ok that's fine. I'll remove the other unused `_category` parameters as well.

> I think we can just publish measureme on crates.io and have lots of major version changes. The other option would be a git dependency. But that's something we should discuss with wg-meta.

I'd personally vote for publishing on crates.io

**Comment by wesleywiser** [comment]
Results of testing on Windows:

| Test Case | Time Taken | 
| ---------- | ----------- |
| regex rustc | 5.02s |
| regex rustc + `-Z self-profile` `FileSerializationSink` | 5.92s |
| regex rustc + `-Z self-profile` `MmapSerializationSink` | 6.04s |

**Comment by michaelwoerister** [comment]
Great, thanks for getting those numbers. Let's use the `FileSerializationSink` on Windows then.

**Comment by michaelwoerister** [comment]
> Eventually I'd like to have query names instead of dep-node names here but I don't know how to exactly do that yet.

It looks like a recent PR fixed this for us by making `DepKind` variants and query names match up `:)`

**Comment by wesleywiser** [comment]
Oh, wonderful! I can make that change tonight. Do we have consensus on how to add the `measureme` dependency? I think that's the last thing that needs a fixup before this can be merged.

**Comment by michaelwoerister** [comment]
Let's go via `crates.io`, I'd say.

**Comment by michaelwoerister** [comment]
Note that `Q::to_dep_node()` is potentially expensive and in non-incremental builds it's not called at all, so we'll have to cache the query name `StringId` afterall. Ideally it would be created at the same time as the query caches and stored in a field somewhere.

**Comment by wesleywiser** [comment]
@michaelwoerister Do you like the approach in edff02e better?

**Comment by michaelwoerister** [comment]
I do -- but it still has some runtime overhead for cloning the query key. Could you try to add an associated constant to `QueryConfig` that holds the right `QueryName` variant? Then we could pass the `QueryName` directly to the `record` methods. You could maybe even replace the current `NAME` constant (not sure if it's used anywhere else).

**Comment by wesleywiser** [comment]
Oh yeah, that's a better idea

File: src/librustc/session/mod.rs [hunk 2]
```diff
@@ -1138,7 +1134,11 @@ fn build_session_(
     driver_lint_caps: FxHashMap<lint::LintId, lint::Level>,
 ) -> Session {
     let self_profiler =
-        if sopts.debugging_opts.self_profile { Some(Arc::new(PlMutex::new(SelfProfiler::new()))) }
+        if sopts.debugging_opts.self_profile {
+            let profiler = SelfProfiler::new();
```
Comment (michaelwoerister) [inline src/librustc/session/mod.rs]: Once this uses proper error handling, I think failing to create the profiler should not be a hard error. Emitting a warning and setting `self_profiler` to `None` should provide the better user experience.

File: src/librustc/util/profiling.rs [hunk 3]
```diff
@@ -1,14 +1,19 @@
...
-use std::time::{Duration, Instant, SystemTime};
+use std::u32;
 
-use crate::session::config::Options;
+use crate::ty::query::QueryName;
 
-use rustc_data_structures::fx::FxHashMap;
+use measureme::{StringId, TimestampKind};
+
+///MmapSerializatioSink is faster on macOS and Linux
```
Comment (michaelwoerister) [inline src/librustc/util/profiling.rs]: Spaces missing after `///`

File: src/librustc/util/profiling.rs [hunk 4]
```diff
@@ -35,409 +40,125 @@ pub enum ProfilerEvent {
...
+            generic_activity_event_kind,
+            incremental_load_result_event_kind,
+            query_blocked_event_kind,
+            query_cache_hit_event_kind,
+        }
+    }
+
+    fn get_query_name_string_id(query_name: QueryName) -> StringId {
+        let discriminant = unsafe {
+            mem::transmute::<_, u64>(mem::discriminant(&query_name))
```
Comment (michaelwoerister) [inline src/librustc/util/profiling.rs]: Maybe make this `mem::transmute::<Discriminant<QueryName>, u64>(mem::discriminant(&query_name))` in order for it to be a little more safe.

**Comment by michaelwoerister** [comment]
I left a few comments. Looks like this is almost ready to merge.

**Comment by wesleywiser** [comment]
Thanks for the feedback @michaelwoerister! I've made those changes, pushed, and marked this PR as ready for review.

File: src/tools/tidy/src/deps.rs [hunk 5]
```diff
@@ -96,6 +96,7 @@ const WHITELIST: &[Crate<'_>] = &[
     Crate("lock_api"),
     Crate("log"),
     Crate("log_settings"),
+    Crate("measureme"),
```
Comment (wesleywiser) [inline src/tools/tidy/src/deps.rs]: I don't know if there's an additional approval process for crate additions but I wanted to call this out just to be safe.

File: src/librustc/session/mod.rs [hunk 6]
```diff
@@ -1137,9 +1133,20 @@ fn build_session_(
...
+    let (self_profiler, profiler_warning) =
+        if sopts.debugging_opts.self_profile {
+            let profiler = SelfProfiler::new();
+            match profiler {
+                Ok(profiler) => {
+                    crate::ty::query::QueryName::register_with_profiler(&profiler);
+                    (Some(Arc::new(profiler)), None)
+                },
+                Err(e) => {
+                    (None, Some(format!("failed to create profiler: {}", e)))
```
Comment (michaelwoerister) [inline src/librustc/session/mod.rs]: You should be able to emit warning already here via:
```rust
early_warn(sopts.error_format, &format!("failed to create profiler: {}", e));
```
That should allow to make the control flow simpler.

File: src/tools/tidy/src/deps.rs (hunk 5 above)
Comment (michaelwoerister) [reply src/tools/tidy/src/deps.rs]: I think that's OK since we already announced that we're going to add this crate.

**Review by michaelwoerister — APPROVED** [review]
Looks great! I left one comment where things could be simplified. r=me once that is resolved.

**Comment by wesleywiser** [comment]
Thanks @michaelwoerister!

@bors r=michaelwoerister

**Comment by michaelwoerister** [comment]
Regarding the category data: We might want to store the mapping of query-name to category in with the profiling data (as part of https://github.com/rust-lang/measureme/issues/15). That way post-processing tools have it available. That would be something for a later PR though.

**Pull request merged**